binary_search_tree/tree.py

```diff
@@ -9,51 +9,148 @@ def __init__(self, key, val = None):
         self.right = None
         
 
-
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if not self.root:
+            self.root = new_node
+            return
+        self.add_helper(self.root, new_node)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def find(self, key):
-        pass
+        if not self.root:
+            return None
+        result = self.find_helper(self.root, key)
+        return result
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
     def inorder(self):
-        pass
+        items = []
+        self.inorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)     
     def postorder(self):
-        pass
+        items = []
+        self.postorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)    
     def height(self):
```

Comment:
✨ however the space complexity would actually be O(n) if the tree is unbalanced and O(log n) if the tree is balanced because of the recursive call stack generated in your helper function!